From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dreeam <61569423+Dreeam-qwq@users.noreply.github.com>
Date: Wed, 27 Nov 2024 23:13:12 -0500
Subject: [PATCH] Only player pushable

Useful for extreme cases like massive entities collide together in a small area

diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 689b44bf0f3a87fb9b2039a65cf90362a0030981..f65cde149ff66db7e2cc761d91b26b26f47874d9 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -3899,7 +3899,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             this.checkAutoSpinAttack(boundingBox, this.getBoundingBox());
         }
 
-        this.pushEntities();
+        if (!org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled) this.pushEntities(); // Leaf - Only player pushable
         profilerFiller.pop();
         // Paper start - Add EntityMoveEvent
         // Purpur start - Ridables
@@ -4052,7 +4052,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
             return;
         }
         // Paper end - don't run getEntities if we're not going to use its result
-        List<Entity> pushableEntities = this.level().getPushableEntities(this, this.getBoundingBox());
+        // Leaf start - Only player pushable
+        final AABB box = this.getBoundingBox();
+        List<Entity> pushableEntities = org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled
+            ? getNearbyPushablePlayers(this, box, EntitySelector.pushableBy(this))
+            : this.level().getPushableEntities(this, box);
+        // Leaf end - Only player pushable
         if (!pushableEntities.isEmpty()) {
             if (this.level() instanceof ServerLevel serverLevel) {
                 // Paper - don't run getEntities if we're not going to use its result; moved up
@@ -4086,6 +4091,44 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         }
     }
 
+    // Leaf start - Only player pushable
+    public List<Entity> getNearbyPushablePlayers(Entity entity, AABB box, Predicate<Entity> conditions) {
+        final Vec3 vec = entity.position;
+        final net.minecraft.core.BlockPos.MutableBlockPos mutablePos = new net.minecraft.core.BlockPos.MutableBlockPos();
+
+        mutablePos.set(vec.x, vec.y, vec.z);
+
+        final ca.spottedleaf.moonrise.common.list.ReferenceList<ServerPlayer> players = ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel) this.level()).moonrise$getNearbyPlayers().getPlayers(
+            mutablePos, ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.SPAWN_RANGE
+        );
+
+        if (players == null) {
+            return new java.util.ArrayList<>();
+        }
+
+        List<Entity> ret = null;
+
+        final ServerPlayer[] raw = players.getRawDataUnchecked();
+        final int len = players.size();
+
+        java.util.Objects.checkFromIndexSize(0, len, raw.length);
+
+        for (int i = 0; i < len; ++i) {
+            final ServerPlayer player = raw[i];
+            if (player != entity && box.intersects(player.getBoundingBox()) && conditions.test(player)) {
+                if (ret == null) {
+                    ret = new java.util.ArrayList<>(len - i);
+                    ret.add(player);
+                } else {
+                    ret.add(player);
+                }
+            }
+        }
+
+        return ret == null ? new java.util.ArrayList<>() : ret;
+    }
+    // Leaf end - Only player pushable
+
     protected void checkAutoSpinAttack(AABB boundingBoxBeforeSpin, AABB boundingBoxAfterSpin) {
         AABB aabb = boundingBoxBeforeSpin.minmax(boundingBoxAfterSpin);
         List<Entity> entities = this.level().getEntities(this, aabb);
diff --git a/net/minecraft/world/entity/decoration/ArmorStand.java b/net/minecraft/world/entity/decoration/ArmorStand.java
index 4f4abeed259a736a08444be12f6e39e057e33315..9d5b4309884f9d5fa42cfd46f96502b1bb7a4dc8 100644
--- a/net/minecraft/world/entity/decoration/ArmorStand.java
+++ b/net/minecraft/world/entity/decoration/ArmorStand.java
@@ -206,7 +206,7 @@ public class ArmorStand extends LivingEntity {
 
     @Override
     protected void pushEntities() {
-        if (!this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups
+        if (org.dreeam.leaf.config.modules.gameplay.OnlyPlayerPushable.enabled || !this.level().paperConfig().entities.armorStands.doCollisionEntityLookups) return; // Paper - Option to prevent armor stands from doing entity lookups // Leaf - Only player pushable
         for (Entity entity : this.level().getEntitiesOfClass(AbstractMinecart.class, this.getBoundingBox(), RIDABLE_MINECARTS)) { // Paper - optimise collisions
             if (this.distanceToSqr(entity) <= 0.2) {
                 entity.push(this);
